### ISSUE TITLE
building 2.1.0a1 fails with python3.6 (gmp-6.1.2/mpfr3.1.6/mpc-1.0.3)

### ISSUE THREAD

**Issue by kiwifb** [aleaxit/gmpy#175]
And gcc-7.2.0 for good measure. I am trying to update the package in Gentoo since it seems sage is currently targeting this release for inclusion as a standard package (because of pplpy). Anyway, python 2.7 compilation goes fine but not 3.6 (haven't tried 3.5 or 3.4)
```
 * python2_7: running distutils-r1_run_phase distutils-r1_python_compile
/usr/bin/python2.7 setup.py build
running build
running build_ext
building 'gmpy2' extension
creating /dev/shm/portage/dev-python/gmpy-2.1.0_alpha1/work/gmpy2-2.1.0a1-python2_7/temp.linux-x86_64-2.7
creating /dev/shm/portage/dev-python/gmpy-2.1.0_alpha1/work/gmpy2-2.1.0a1-python2_7/temp.linux-x86_64-2.7/src
x86_64-pc-linux-gnu-gcc -march=native -O2 -pipe -fPIC -DSHARED=1 -I./src -I/usr/include/python2.7 -c src/gmpy2.c -o /dev/shm/portage/dev-python/gmpy-2.1.0_alpha1/work/gmpy2-2.1.0a1-python2_7/temp.linux-x86_64-2.7/src/gmpy2.o
x86_64-pc-linux-gnu-gcc -shared -Wl,-O1 -Wl,--as-needed -march=native -O2 -pipe /dev/shm/portage/dev-python/gmpy-2.1.0_alpha1/work/gmpy2-2.1.0a1-python2_7/temp.linux-x86_64-2.7/src/gmpy2.o -L/usr/lib64 -lgmp -lmpfr -lmpc -lpython2.7 -o /dev/shm/portage/dev-python/gmpy-2.1.0_alpha1/work/gmpy2-2.1.0a1-python2_7/lib/gmpy2.so
 * python3_6: running distutils-r1_run_phase distutils-r1_python_compile
/usr/bin/python3.6 setup.py build
running build
running build_ext
building 'gmpy2' extension
creating /dev/shm/portage/dev-python/gmpy-2.1.0_alpha1/work/gmpy2-2.1.0a1-python3_6/temp.linux-x86_64-3.6
creating /dev/shm/portage/dev-python/gmpy-2.1.0_alpha1/work/gmpy2-2.1.0a1-python3_6/temp.linux-x86_64-3.6/src
x86_64-pc-linux-gnu-gcc -march=native -O2 -pipe -fPIC -DSHARED=1 -I./src -I/usr/include/python3.6m -c src/gmpy2.c -o /dev/shm/portage/dev-python/gmpy-2.1.0_alpha1/work/gmpy2-2.1.0a1-python3_6/temp.linux-x86_64-3.6/src/gmpy2.o
In file included from src/gmpy2.c:400:0:
src/gmpy2_convert_gmp.c: In function ‘GMPy_PyLong_From_MPZ’:
src/gmpy2_convert_gmp.c:197:32: error: ‘SHIFT’ undeclared (first use in this function); did you mean ‘S_IFMT’?
     assert(mp_bits_per_limb >= SHIFT);
                                ^
src/gmpy2_convert_gmp.c:197:32: note: each undeclared identifier is reported only once for each function it appears in
error: command 'x86_64-pc-linux-gnu-gcc' failed with exit status 1
```
I see `SHIFT` being defined in `/usr/include/python2.7/longintrepr.h` but nowhere in python3.6. Yet  I see travis targeting python3.6. What am I missing?

**Comment by kiwifb** [comment]
I'll note that in python-2.7, `SHIFT` is an alias for `PyLong_SHIFT` and that constant is defined (with the same value) for both python-2.7 and 3.6.

**Comment by isuruf** [comment]
Difference between travis-ci and yours is that `-DNDEBUG` is not used in your build. Adding `-DNDEBUG` would remove the asserts when the pre-processor is run.

```
In [181]: import sysconfig
     ...: print(sysconfig.get_config_vars()['PY_CFLAGS'])
     ...: 
-Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -Wstrict-prototypes
```

**Comment by kiwifb** [comment]
Thanks for the explanation @isuruf. Still, I think it should be safe - and the right thing - to just "promote" that `SHIFT` to `PyLong_SHIFT`.
After doing that I am working on getting the testsuite to work on gentoo. Looks like it will be fun.

**Comment by isuruf** [comment]
> Still, I think it should be safe - and the right thing - to just "promote" that SHIFT to PyLong_SHIFT.

Sure. Btw, a Release build shouldn't really run asserts. Hurts performance.

**Comment by kiwifb** [comment]
Cannot argue with that. It is annoying to have to do it manually as a packager but this is "alpha1" so I should let it slide and just direct my "unhappy thoughts" to the people who insist on using alpha stuff for their own stable releases. As sage-on-gentoo maintainer, over the years, I got several versions of that T-shirt (as in "been there, done that, got the T-shirt").

**Comment by isuruf** [comment]
IMO, this is not gmpy's fault. gmpy uses distutils which reuses the flags that python was built with. python on gentoo doesn't have that flag and therefore it's not passed on to the extensions. This affects all python extensions, not only gmpy

**Comment by kiwifb** [comment]
We are straying away, but you certainly have an interesting point. I guess this is a consequence of how gentoo is setup and what it does. In gentoo it is up to the individual package, if possible, and with the user in the driver seat to turn it on or off - sanity not included :) . My own feeling here is, this is alpha, debug is supposed to be on in alpha.

**Comment by casevh** [comment]
It is a bug in gmpy2. The assert statement should be:

assert(mp_bits_per_limb >= PyLong_SHIFT);

The use of SHIFT was a leftover from older Python versions and I missed it in the cleanup. And I missed testing the DEBUG path when I stopped testing with Python 3.3. (Long story...)

I plan to release alpha2 after I fix the warnings with MPFR 4.0.0.

**Comment by videlec** [comment]
This has been fixed in `89e5d4e9`. Could the issue be closed!?

**Comment by kiwifb** [comment]
Sure!
